### PR TITLE
Reimplement Map Markers from BWMF

### PR DESCRIPTION
**Pull request by MikeMatrix**
[WIP]

- [x] Implement method to load configuration from any configuration File and save it to memory
- [x] Implement render method for Groups/Units
- [x] Implement render of Group Members
- [ ] Implement convenience methods to change settings on the fly?
- [ ] Implement a JIP safe way to change settings in runtime
- [ ] Integrate changes into bwmf when this becomes an active part of the modpack

This also introduces a map datastructure to the codebase as a side effect, might be worth investigating, if changes could be made to the way CBA does its "hashes".

Current configuration format will be along the lines of
```
class POTATO_mapMarkers_Configuration {
    // General Configuration
    enabled = 1;
    updateRate = 5;

    // Which factions can see what other factions. They are defined as classes, because of potential black/whitelisting down the line, that can be defined there
    class Visibility {
        class blu_f {
            class blu_f;
        };
        class ind_f {
            class ind_f;
        };
        class opf_f {
            class opf_f;
        };
        class rhs_faction_msv {
            class rhs_faction_msv;
            class rhs_faction_vvs;
        };
        class rhs_faction_vvs {
            class rhs_faction_msv;
            class rhs_faction_vvs;
        };
    };

    // Specific units can have attached markers. Units will be looked up by their assigned variable name
    class Units {
        // Inheritable Bases (these will be moved to a separate included file in the framework, they will be able to be used for both Units and Groups)
        // This also means that in case a unit/group has this specific name, they will get this marker configuration attached
        // This is a tradeoff of having inheritance, but still loading all data into memory. I might introduce a specific prefix to not load a class into memory and only use it for inheritance
        class Medic { 
            icon = "\A3\ui_f\data\map\markers\nato\b_med.paa";
            size[] = {12, 12};
            color[] = {1, 0.753, 0.796, 1};
            displayText = 1; // This might become optional as most of these, considering I can default them in code
        };
        class RedFTL {
            icon = "\A3\ui_f\data\map\markers\nato\b_inf.paa";
            size[] = {24, 24};
            color[] = {1, 0, 0, 1};
            displayText = 1;
        };
        class BlueFTL {
            icon = "\A3\ui_f\data\map\markers\nato\b_inf.paa";
            size[] = {24, 24};
            color[] = {0, 0, 1, 1};
            displayText = 1;
        };
        class GreenFTL {
            icon = "\A3\ui_f\data\map\markers\nato\b_inf.paa";
            size[] = {24, 24};
            color[] = {0, 1, 0, 1};
            displayText = 1;
        };

        class UnitMSV_ASL_SR : RedFTL {};
        class UnitMSV_BSL_SR : BlueFTL {};
        class UnitMSV_CSL_SR : GreenFTL {};
        class UnitMSV_DSL_SR : RedFTL {};
        class UnitMSV_ESL_SR : BlueFTL {};
        class UnitMSV_FSL_SR : GreenFTL {};
    };
    
    class Groups {
        class NATO_COY { // If group names contain spaces, they will likely be replaced by underscores and looked up that way
            icon = "\A3\ui_f\data\map\markers\nato\b_hq.paa";
            color[] = {1, 1, 0, 1};
            size[] = {24, 24};
            displayText = 1;
        };
    };
};
```

This system should be designed in a way, that will allow for the use of EDEN based value configuration without too much effort.

### CODE REVIEW

**Comment by MikeMatrix** [comment]
Tagging @lavinrp for his Mission Making tool, as it's relevant.

File: addons/datastructures/functions/fnc_map_create.sqf [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+if (_pairs isEqualTo []) exitWith {DEFAULT_STRUCTURE};
+
+private _result = DEFAULT_STRUCTURE;
+
+_pairs params ["_keys", "_values"];
+
+{
+    _x params ["_key", "_value"];
+    [_result, _key, _value] call FUNC(map_put);
+} forEach _pairs;
```
Comment (AACO) [inline addons/datastructures/functions/fnc_map_create.sqf]: Should this be a count?

File: addons/datastructures/script_component.hpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#define COMPONENT datastructures
+#include "\z\potato\addons\core\script_mod.hpp"
+
+#define DEBUG_MODE_FULL
+#define DISABLE_COMPILE_CACHE
+#define CBA_DEBUG_SYNCHRONOUS
+
+#ifdef DEBUG_ENABLED_SERVERMONITOR
```
Comment (AACO) [inline addons/datastructures/script_component.hpp]: Shouldn't these be more specific debug defines?

File: addons/mapMarkers/functions/fnc_drawMap.sqf [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        ["_textLeft", "", [""]]
+    ];
+    _size params ["_width", "_height"];
+    _width = _width * _sizeFactor;
+    _height = _height * _sizeFactor;
+    private _textSize = if ((_mapScale * _mapSize) >= 0.005) then {0} else {0.02 * _sizeFactor};
+
+    _mapControl drawIcon [_icon, _color, _position, _width, _height, _direction, _textLeft, 1, _textSize, "TahomaB", "left"];
+    nil
+} count GVAR(groupMemberDrawCache);
```
Comment (AACO) [inline addons/mapMarkers/functions/fnc_drawMap.sqf]: (question) will these caches be updated on JIPs?

File: addons/mapMarkers/XEH_postInit.sqf [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+// Load configurations from the missions config file
+[missionConfigFile] call FUNC(loadFromConfigFile);
+
+// Start drawCache update using the provided updateInterval
+[{[] call FUNC(updateMarkerDrawCache)}, GVAR(updateInterval), []] call CBA_fnc_addPerFrameHandler;
+[] call FUNC(updateMarkerDrawCache);
+
+// Start groupMemberDrawCache update
+[{[] call FUNC(updateGroupMemberDrawCache)}, 0, []] call CBA_fnc_addPerFrameHandler;
```
Comment (AACO) [inline addons/mapMarkers/XEH_postInit.sqf]: Add way to disable these intra-fireteam markers?

File: addons/mapMarkers/XEH_postInit.sqf [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#include "script_component.hpp"
+
+// Load configurations from the missions config file
+[missionConfigFile] call FUNC(loadFromConfigFile);
+
+// Start drawCache update using the provided updateInterval
+[{[] call FUNC(updateMarkerDrawCache)}, GVAR(updateInterval), []] call CBA_fnc_addPerFrameHandler;
```
Comment (AACO) [inline addons/mapMarkers/XEH_postInit.sqf]: Do we want to update the cache every interval? Maybe only update the cache on player init event?

**Comment by AACO** [comment]
Moving in a 3DEN direction (sorry totally not stale @MikeMatrix PR from January)

**Pull request closed without merging**